File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.java
package ithaca_transit.android.cornellappdev.com.ithaca_transit.Models;


import java.util.Date;

public class Direction {

    private static final Date BLANK_TIME = new Date();

    private int delay;
    private Double distance;
    private Coordinate endLocation;
    private Date endTime;
    private String name;
    private Coordinate[] path;
    private int routeNumber;
    private Coordinate startLocation;
    private Date startTime;
    private Boolean stayOnBusForTransfer;
    private Place[] stops;
    private String[] tripIdentifiers;
    private String type;

    public Direction(int delay, Double distance,
            Coordinate endLocation, Date endTime, String name,
            Coordinate[] path, int routeNumber,
            Coordinate startLocation, Date startTime, Boolean stayOnBusForTransfer,
            Place[] stops, String[] tripIdentifiers, String type) {
        this.delay = delay;
        this.distance = distance;
        this.endLocation = endLocation;
        this.endTime = endTime;
        this.name = name;
        this.path = path;
        this.routeNumber = routeNumber;
        this.startLocation = startLocation;
        this.startTime = startTime;
        this.stayOnBusForTransfer = stayOnBusForTransfer;
        this.stops = stops;
        this.tripIdentifiers = tripIdentifiers;
        this.type = type;
    }

    public Direction(Double distance, String name, String type) {
        this.distance = distance;
        this.name = name;
        this.type = type;
    }

    public String locationDescription() {
        if (type != null) {
            switch (type) {
                case "walk":
                    return "Walk to " + this.name;
                case "arrive":
                    return "Get off at " + this.name;
                case "depart":
                    return "at " + this.name;
                case "transfer":
                    return "at " + this.name + ". Stay on bus";
            }
        }
        return "";
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public Integer getDelay() {
        return delay;
    }

    public Double getDistance() {
        return distance;
    }

    public Coordinate getEndLocation() {
        return endLocation;
    }

    public Date getEndTime() {
        return endTime;
    }

    public Coordinate[] getPath() {
        return path;
    }

    public Integer getRouteNumber() {
        return routeNumber;
    }

    public Coordinate getStartLocation() {
        return startLocation;
    }

    public Date getStartTime() {
        return startTime;
    }

    public Boolean getStayOnBusForTransfer() {
        return stayOnBusForTransfer;
    }

    public Place[] getStops() {
        return stops;
    }

    public String[] getTripIdentifiers() {
        return tripIdentifiers;
    }

    public String getName() {
        return name;
    }

    public void setDistance(Double distance) {
        this.distance = distance;
    }

    /* Delete last element of stops array, as we want to depict the last stop as a direction in detail view
     * For example, if the last stop is Duffield, we want to show "Get off at Duffield" instead of
     * Duffield only showing up as a stop in the detail view
     */
    public void updateStops(){
        int numStops = stops.length;
        Place[] newStops = new Place[numStops-1];

        for(int count=0;count<numStops-1;count++){
            newStops[count] = stops[count];
        }
        stops = newStops;
    }
}

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt
package ithaca_transit.android.cornellappdev.com.ithaca_transit.Models

import com.google.android.gms.maps.model.LatLng
import java.util.Date

enum class DirectionType {
    ARRIVE, DEPART, TRANSFER, WALK
}

class Direction(delay: Int, endLocation: LocationObject, endTime: Date,
                name: String, path: Array<LatLng>, routeNumber: Int, startLocation: LocationObject,
                startTime: Date, stayOnBusForTransfer: Boolean, stops: Array<LocationObject>,
                travelDistance: Double, tripIdentifiers: Array<String>, type: DirectionType) {

    private var delay: Int? = null
    private var endLocation: LocationObject? = null
    private var endTime: Date? = null
    private var name: String? = null
    private var path: Array<LatLng>? = null
    private var routeNumber: Int? = null
    private var startLocation: LocationObject? = null
    private var startTime: Date? = null
    private var stayOnBusForTransfer: Boolean? = null
    private var stops: Array<LocationObject>? = null
    private var travelDistance: Double? = null
    private var tripIdentifiers: Array<String>? = null
    private var type: DirectionType? = null


    init {
        this.delay = delay
        this.endLocation = endLocation
        this.endTime = endTime
        this.name = name
        this.path = path
        this.routeNumber = routeNumber
        this.startLocation = startLocation
        this.startTime = startTime
        this.stayOnBusForTransfer = stayOnBusForTransfer
        this.stops = stops
        this.travelDistance = travelDistance
        this.tripIdentifiers = tripIdentifiers
        this.type = type
    }

    //static variables
    companion object {
        val blankLocation = LocationObject()
        val blankTime = Date()
    }

    constructor(name: String) : this(0, blankLocation, blankTime, if (name != null) name else "",
            emptyArray<LatLng>(), 0, blankLocation, blankTime, false,
            emptyArray<LocationObject>(), 0.0, emptyArray<String>(), DirectionType.ARRIVE)

    fun locationDescription(): String {
        var instruct = ""
        when (type) {
            DirectionType.WALK -> instruct = "Walk to " + name
            DirectionType.ARRIVE -> instruct = "Get off at " + name
            DirectionType.DEPART -> instruct = "at " + name
            DirectionType.TRANSFER -> instruct = "at " + name + ". Stay on bus"
        }
        return instruct
    }

}